File: parse_script.js
const fs = require('fs');
const path = require('path');

const scriptPath = path.join(__dirname, 'GAL劇本_椎名日和線.md');
const outputPath = path.join(__dirname, 'game_script.json');

function parseScript() {
  if (!fs.existsSync(scriptPath)) {
    console.error(`Script file not found: ${scriptPath}`);
    process.exit(1);
  }

  const rawContent = fs.readFileSync(scriptPath, 'utf8');
  const lines = rawContent.split(/\r?\n/);

  const game = {
    acts: []
  };

  let currentAct = null;
  let currentSpeaker = null;
  let currentExpression = null;
  let currentChoice = null;
  let currentBranch = null;

  // Regex definitions
  const actRegex = /^##\s+第\s*(\d+)\s*幕：\s*(.*)$/;
  const locRegex = /^\*\*地點：\s*(.*)\*\*$/;
  const timeRegex = /^\*\*時間：\s*(.*)\*\*$/;
  const bgmRegex = /^\*\*BGM：\s*(.*)\*\*$/;
  const seRegex = /^\*\*(特殊效果|特殊特效)：\s*(.*)\*\*$/;
  const sceneDescRegex = /^\*\*【場景：\s*(.*)】\*\*$/;

  // Speaker header e.g. **清隆**（独白）： or **日和**（通常·抱書，未注意到清隆）： or **班導**：
  const speakerRegex = /^\*\*(.*?)\*\*(?:（(.*?)）)?：$/;

  // Dialogue quote e.g. > 「......」 or > （......）
  const dialogueRegex = /^>\s*(.*)$/;

  // Choice option e.g. > **A）「......」**
  const choiceOptionRegex = /^>\s*\*\*([A-C])）\s*(.*)\*\*$/;

  // Choice block header e.g. **【選擇肢 1-1】**
  const choiceHeaderRegex = /^\*\*【選擇肢\s*(\d+-\d+)】\*\*$/;

  // Choice branch header e.g. **【選擇A】** or **【選擇A → 進入禮物劇情】**
  const branchHeaderRegex = /^\*\*【選擇([A-C])(?:.*?)\】\*\*$/;

  // CG unlock e.g. **【獲得：CG-01「圖書館的午後」— 日和靠窗閱讀的側臉，陽光灑在髮絲上】**
  const cgRegex = /^\*\*【獲得：\s*(CG-\d+)「(.*?)」\s*—\s*(.*?)】\*\*$/;

  // Act end e.g. **【幕結束】**
  const actEndRegex = /^\*\*【幕結束】\*\*$/;

  // Convergence e.g. **【匯流：以下劇情三選共通】**
  const convergenceRegex = /^\*\*【匯流：\s*(.*)】\*\*$/;

  // Action e.g. *——清隆輕輕走過去，在附近的書架前停下，抽出一本書。*
  const actionRegex = /^\*(.*?)\*$/;

  function addNode(node) {
    if (!currentAct) return;
    if (currentChoice && currentBranch) {
      // Find the option in the current choice
      const option = currentChoice.options.find(opt => opt.key === currentBranch);
      if (option) {
        option.nodes.push(node);
      } else {
        console.warn(`Warning: Branch option ${currentBranch} not found in choice ${currentChoice.id}`);
      }
    } else {
      currentAct.nodes.push(node);
    }
  }

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;

    // 1. Act Header
    let match = line.match(actRegex);
    if (match) {
      // If there was an active choice, close it before moving to the next act
      if (currentChoice) {
        currentAct.nodes.push(currentChoice);
        currentChoice = null;
        currentBranch = null;
      }

      currentAct = {
        id: parseInt(match[1], 10),
        title: match[2].trim(),
        metadata: {
          location: '',
          time: '',
          bgm: '',
          se: ''
        },
        nodes: []
      };
      game.acts.push(currentAct);
      currentSpeaker = null;
      currentExpression = null;
      continue;
    }

    if (!currentAct) continue;

    // 2. Metadata
    if ((match = line.match(locRegex))) {
      currentAct.metadata.location = match[1].trim();
      continue;
    }
    if ((match = line.match(timeRegex))) {
      currentAct.metadata.time = match[1].trim();
      continue;
    }
    if ((match = line.match(bgmRegex))) {
      currentAct.metadata.bgm = match[1].trim();
      continue;
    }
    if ((match = line.match(seRegex))) {
      currentAct.metadata.se = match[2].trim();
      continue;
    }

    // 3. Act End
    if (line.match(actEndRegex)) {
      if (currentChoice) {
        currentAct.nodes.push(currentChoice);
        currentChoice = null;
        currentBranch = null;
      }
      addNode({
        type: 'act_end'
      });
      continue;
    }

    // 4. CG Unlock
    if ((match = line.match(cgRegex))) {
      addNode({
        type: 'cg_unlock',
        cgId: match[1].trim(),
        title: match[2].trim(),
        description: match[3].trim()
      });
      continue;
    }

    // 5. Choice block start
    if ((match = line.match(choiceHeaderRegex))) {
      // If there was an active choice, close it first
      if (currentChoice) {
        currentAct.nodes.push(currentChoice);
      }
      currentChoice = {
        type: 'choice',
        id: match[1],
        options: []
      };
      currentBranch = null;
      continue;
    }

    // 6. Choice options (lines starting with > **A）...)
    if ((match = line.match(choiceOptionRegex))) {
      if (currentChoice) {
        currentChoice.options.push({
          key: match[1],
          text: match[2].trim(),
          nodes: []
        });
      } else {
        console.warn(`Warning: Choice option found outside of a choice block: ${line}`);
      }
      continue;
    }

    // 7. Choice Branch Start (e.g. **【選擇A】**)
    if ((match = line.match(branchHeaderRegex))) {
      currentBranch = match[1];
      continue;
    }

    // 8. Convergence Point (e.g. **【匯流：...】**)
    if ((match = line.match(convergenceRegex))) {
      if (currentChoice) {
        currentAct.nodes.push(currentChoice);
        currentChoice = null;
        currentBranch = null;
      }
      addNode({
        type: 'convergence',
        text: match[1].trim()
      });
      continue;
    }

    // 9. Scene Description (e.g. **【場景：...】**)
    if ((match = line.match(sceneDescRegex))) {
      // A scene description can also implicitly end a choice block if we are in one and hit a convergence scene
      if (currentChoice && currentBranch) {
        currentAct.nodes.push(currentChoice);
        currentChoice = null;
        currentBranch = null;
      }
      addNode({
        type: 'scene_desc',
        text: match[1].trim()
      });
      continue;
    }

    // 10. Speaker Header (e.g. **清隆**（独白）：)
    if ((match = line.match(speakerRegex))) {
      currentSpeaker = match[1].trim();
      currentExpression = match[2] ? match[2].trim() : null;
      continue;
    }

    // 11. Dialogue Quote (e.g. > 「......」 or > （......）)
    if ((match = line.match(dialogueRegex))) {
      if (currentSpeaker) {
        const textVal = match[1].trim();
        const isMonologue = textVal.startsWith('（') && textVal.endsWith('）');
        // Clean outer brackets if it's monologue/dialogue
        let cleanedText = textVal;
        if (isMonologue) {
          cleanedText = textVal.substring(1, textVal.length - 1);
        } else if (textVal.startsWith('「') && textVal.endsWith('」')) {
          cleanedText = textVal.substring(1, textVal.length - 1);
        }

        addNode({
          type: 'dialogue',
          speaker: currentSpeaker,
          expression: currentExpression,
          text: cleanedText,
          isMonologue: isMonologue
        });
        // We do not reset speaker/expression because they might speak multiple lines,
        // although in the script each paragraph has a speaker header.
      } else {
        // It's a dialogue quote without speaker, could be narrator or system text
        const textVal = match[1].trim();
        addNode({
          type: 'narrative',
          text: textVal
        });
      }
      continue;
    }

    // 12. Action lines (e.g. *——清隆輕輕走過去...*)
    if ((match = line.match(actionRegex))) {
      let actionText = match[1].trim();
      if (actionText.startsWith('——')) {
        actionText = actionText.substring(2).trim();
      }
      addNode({
        type: 'action',
        text: actionText
      });
      continue;
    }

    // If it's a separator line like --- and we are in a choice branch,
    // we don't necessarily end the branch yet (until we see another branch or convergence).
    if (line === '---') {
      continue;
    }

    // If we get here, it's an unhandled line format (could be headers, notes, theme song concept, etc.)
    // We can just log it or skip it
    // console.log(`Skipped line: ${line}`);
  }

  // Final check: if game ended but choice is still open, push it
  if (currentChoice) {
    currentAct.nodes.push(currentChoice);
  }

  fs.writeFileSync(outputPath, JSON.stringify(game, null, 2), 'utf8');
  console.log(`Successfully parsed script! Saved to: ${outputPath}`);
  console.log(`Total acts parsed: ${game.acts.length}`);
}

parseScript();

File: parse_script.py
import os
import re
import json

def parse_script():
    script_path = os.path.join(os.path.dirname(__file__), 'GAL劇本_椎名日和線.md')
    output_path = os.path.join(os.path.dirname(__file__), 'game_script.json')

    if not os.path.exists(script_path):
        print(f"Script file not found: {script_path}")
        return

    with open(script_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    game = {
        "acts": []
    }

    current_act = None
    current_speaker = None
    current_expression = None
    current_choice = None
    current_branch = None

    # Regex definitions
    act_regex = re.compile(r'^##\s+第\s*(\d+)\s*幕：\s*(.*)$')
    loc_regex = re.compile(r'^\*\*地點：\s*(.*)\*\*$')
    time_regex = re.compile(r'^\*\*時間：\s*(.*)\*\*$')
    bgm_regex = re.compile(r'^\*\*BGM：\s*(.*)\*\*$')
    se_regex = re.compile(r'^\*\*(特殊效果|特殊特效)：\s*(.*)\*\*$')
    scene_desc_regex = re.compile(r'^\*\*【場景：\s*(.*)】\*\*$')
    speaker_regex = re.compile(r'^\*\*(.*?)\*\*(?:（(.*?)）)?：$')
    dialogue_regex = re.compile(r'^>\s*(.*)$')
    choice_option_regex = re.compile(r'^>\s*\*\*([A-C])）\s*(.*)\*\*$')
    choice_header_regex = re.compile(r'^\*\*【選擇肢\s*(\d+-\d+)】\*\*$')
    branch_header_regex = re.compile(r'^\*\*【選擇([A-C])(?:.*?)\】\*\*$')
    cg_regex = re.compile(r'^\*\*【獲得：\s*(CG-\d+)「(.*?)」\s*—\s*(.*?)】\*\*$')
    act_end_regex = re.compile(r'^\*\*【幕結束】\*\*$')
    convergence_regex = re.compile(r'^\*\*【匯流：\s*(.*)】\*\*$')
    action_regex = re.compile(r'^\*(.*?)\*$')

    def add_node(node):
        nonlocal current_act, current_choice, current_branch
        if not current_act:
            return
        if current_choice and current_branch:
            # Find the option in the current choice
            option = next((opt for opt in current_choice['options'] if opt['key'] == current_branch), None)
            if option:
                option['nodes'].append(node)
            else:
                print(f"Warning: Branch option {current_branch} not found in choice {current_choice['id']}")
        else:
            current_act['nodes'].append(node)

    for line in lines:
        line = line.strip()
        if not line:
            continue

        # 1. Act Header
        match = act_regex.match(line)
        if match:
            if current_choice:
                current_act['nodes'].append(current_choice)
                current_choice = None
                current_branch = None

            current_act = {
                "id": int(match.group(1)),
                "title": match.group(2).strip(),
                "metadata": {
                    "location": '',
                    "time": '',
                    "bgm": '',
                    "se": ''
                },
                "nodes": []
            }
            game['acts'].append(current_act)
            current_speaker = None
            current_expression = None
            continue

        if not current_act:
            continue

        # 2. Metadata
        match = loc_regex.match(line)
        if match:
            current_act['metadata']['location'] = match.group(1).strip()
            continue
        
        match = time_regex.match(line)
        if match:
            current_act['metadata']['time'] = match.group(1).strip()
            continue
            
        match = bgm_regex.match(line)
        if match:
            current_act['metadata']['bgm'] = match.group(1).strip()
            continue
            
        match = se_regex.match(line)
        if match:
            current_act['metadata']['se'] = match.group(2).strip()
            continue

        # 3. Act End
        if act_end_regex.match(line):
            if current_choice:
                current_act['nodes'].append(current_choice)
                current_choice = None
                current_branch = None
            add_node({"type": 'act_end'})
            continue

        # 4. CG Unlock
        match = cg_regex.match(line)
        if match:
            add_node({
                "type": 'cg_unlock',
                "cgId": match.group(1).strip(),
                "title": match.group(2).strip(),
                "description": match.group(3).strip()
            })
            continue

        # 5. Choice block start
        match = choice_header_regex.match(line)
        if match:
            if current_choice:
                current_act['nodes'].append(current_choice)
            current_choice = {
                "type": 'choice',
                "id": match.group(1),
                "options": []
            }
            current_branch = None
            continue

        # 6. Choice options
        match = choice_option_regex.match(line)
        if match:
            if current_choice:
                current_choice['options'].append({
                    "key": match.group(1),
                    "text": match.group(2).strip(),
                    "nodes": []
                })
            else:
                print(f"Warning: Choice option found outside of a choice block: {line}")
            continue

        # 7. Choice Branch Start
        match = branch_header_regex.match(line)
        if match:
            current_branch = match.group(1)
            continue

        # 8. Convergence Point
        match = convergence_regex.match(line)
        if match:
            if current_choice:
                current_act['nodes'].append(current_choice)
                current_choice = None
                current_branch = None
            add_node({
                "type": 'convergence',
                "text": match.group(1).strip()
            })
            continue

        # 9. Scene Description
        match = scene_desc_regex.match(line)
        if match:
            if current_choice and current_branch:
                current_act['nodes'].append(current_choice)
                current_choice = None
                current_branch = None
            add_node({
                "type": 'scene_desc',
                "text": match.group(1).strip()
            })
            continue

        # 10. Speaker Header
        match = speaker_regex.match(line)
        if match:
            current_speaker = match.group(1).strip()
            current_expression = match.group(2).strip() if match.group(2) else None
            continue

        # 11. Dialogue Quote
        match = dialogue_regex.match(line)
        if match:
            if current_speaker:
                text_val = match.group(1).strip()
                is_monologue = text_val.startswith('（') and text_val.endswith('）')
                cleaned_text = text_val
                if is_monologue:
                    cleaned_text = text_val[1:-1]
                elif text_val.startswith('「') and text_val.endswith('」'):
                    cleaned_text = text_val[1:-1]

                add_node({
                    "type": 'dialogue',
                    "speaker": current_speaker,
                    "expression": current_expression,
                    "text": cleaned_text,
                    "isMonologue": is_monologue
                })
            else:
                text_val = match.group(1).strip()
                add_node({
                    "type": 'narrative',
                    "text": text_val
                })
            continue

        # 12. Action lines
        match = action_regex.match(line)
        if match:
            action_text = match.group(1).strip()
            if action_text.startswith('——'):
                action_text = action_text[2:].strip()
            add_node({
                "type": 'action',
                "text": action_text
            })
            continue

        if line == '---':
            continue

    if current_choice:
        current_act['nodes'].append(current_choice)

    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(game, f, ensure_ascii=False, indent=2)
    
    print(f"Successfully parsed script! Saved to: {output_path}")
    print(f"Total acts parsed: {len(game['acts'])}")

if __name__ == "__main__":
    parse_script()
